pallets/sto/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+            // Tuple of (tier_id, amount to purchase from that tier).
+            let mut purchases = Vec::new();
+
+            for (id, tier) in fundraiser.tiers.iter().enumerate() {
+                // fulfilled the investment amount
+                if remaining == 0.into() {
+                    break
+                }
+
+                // tier is exhausted, move on
+                if tier.remaining == 0.into() {
+                    continue
+                }
+
+                // Check if this tier can fulfil the remaining investment amount.
+                // If it can, purchase the remaining amount.
+                // If it can't, purchase what's remaining in the tier.
+                let purchase_amount = if tier.remaining >= remaining {
+                    remaining
+                } else {
+                    tier.remaining
+                };
+
+                remaining -= purchase_amount;
+                purchases.push((id, purchase_amount));
+                cost = cost.checked_add(
+                    &remaining
+                    .checked_mul(&tier.price)
+                    .ok_or(Error::<T>::Overflow)?
+                ).ok_or(Error::<T>::Overflow)?;
+            }
+
+            ensure!(remaining == 0.into(), Error::<T>::InsufficientTokensRemaining);
+
+            let primary_issuance_agent = T::Asset::primary_issuance_agent(&offering_asset);
+            let legs = vec![
+                Leg {
+                    from: fundraiser.offering_portfolio,
+                    to: investment_portfolio,
+                    asset: offering_asset,
+                    amount: investment_amount
+                },
+                Leg {
+                    from: funding_portfolio,
+                    to: fundraiser.raising_portfolio,
+                    asset: fundraiser.raising_asset,
+                    amount: cost
+                }
+            ];
+
+            with_transaction(|| {
+               let instruction_id = Settlement::<T>::base_add_instruction(
+                    primary_issuance_agent,
+                    fundraiser.venue_id,
+                    SettlementType::SettleOnAuthorization,
+                    None,
+                    legs
+                )?;
+
+                let portfolios = vec![investment_portfolio, funding_portfolio];
+
+                match reciept {
+                    Some(reciept) => Settlement::<T>::authorize_with_receipts(
+                        origin,
+                        instruction_id,
+                        vec![reciept],
+                        portfolios
+                    ).map_err(|e| e.error)?,
+                    None => Settlement::<T>::authorize_instruction(origin, instruction_id, portfolios).map_err(|e| e.error)?,
+                };
+
+                let portfolios= vec![investment_portfolio, funding_portfolio].into_iter().collect::<BTreeSet<_>>();
```

Comment:
```suggestion
                let portfolios= vec![fundraiser.offering_portfolio, fundraiser.raising_portfolio].into_iter().collect::<BTreeSet<_>>();
```

If this didn't get caught by test cases, I think we need better test cases.

Reply:
Made the changes to settlement and this now fails properly.